### PR TITLE
Replace assembly with verified C/C++ replacements.

### PR DESCRIPTION
**Pull request by OmniBlade**
Different approach to ensure C replacements function the same/similarly enough compared to the asm code they replace. Intended to superceed #33.

### CODE REVIEW

**Comment by wdigger** [comment]
Is there any reason to reimplement asm in C files with _cdecl calling convention? There are only a few C files in the project, clearly this is not the main trend.

**Comment by hifi** [comment]
I would actually also want to drop the implicit cdecl because it either needs to be redefined globally for non-Windows builds (`__cdecl` is not defined by gcc on Linux). As far as I understand it is the default convention for plain functions for both C and C++ so it is redundant anyway.

EDIT: Yes I know it's x86 specific but we don't have assembly for other platforms so the calling convention doesn't matter anywhere else.

**Comment by OmniBlade** [comment]
Yeah I can start dropping the __cdecl stuff, it will date back to watcom whee the calling convention was register based by default.

**Comment by OmniBlade** [comment]
Okay, reverted to using embeded data so the tests can be carried out even without being able to build asm on a given platform after discussion with @hifi and a view to future usefulness of the tests. Putting this up for merge now so the testing infrastructure is there to be added to and will add new functions in subsequent PRs.

File: CMakeLists.txt [hunk 1]
```diff
@@ -48,9 +57,13 @@ option(BUILD_REMASTERRA "Build Red Alert remaster dll." ${DEFAULT_REMASTER})
 option(BUILD_VANILLATD "Build Tiberian Dawn executable." ON)
 option(BUILD_VANILLARA "Build Red Alert executable." ON)
 option(MAP_EDITORTD "Include internal scenario editor in Tiberian Dawn build." OFF)
-option(MAP_EDITORRA "Include internal scenario editor in Red Alert build." OFF)
```
Comment (hifi) [inline CMakeLists.txt:51]: Whoops?

**Review by hifi — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (OmniBlade) [reply CMakeLists.txt:51]: Indeed, I'll get that fixed.

**Review by hifi — APPROVED** [review]
Builds and runs, tests pass with and without asm.

**Pull request merged**